pil-analyzer/src/type_inference.rs

```diff
@@ -0,0 +1,675 @@
+                }
+            } else {
+                self.process_expression(value).map(|ty| {
+                    self.inferred_types.insert(name.to_string(), ty);
+                })
+            })
+            .map_err(|e| format!("Error type checking the symbol {name} = {value}:\n{e}",))?;
+        }
+
+        self.check_expressions(expressions)?;
+
+        // From this point on, the substitutions are fixed.
+
+        // Now we check for all symbols that are not declared as a type scheme that they
+        // can resolve to a concrete type.
+        for (name, declared_type) in &self.declared_types {
+            if declared_type.vars.is_empty() {
+                // It is not a type scheme, see if we were able to derive a concrete type.
+                let inferred = self.substitute_to(declared_type.ty.clone());
+                if !inferred.is_concrete_type() {
+                    let inferred_scheme = self.to_type_scheme(inferred);
+                    Err(format!(
+                        "Could not derive a concrete type for symbol {name}.\nInferred type scheme: {}\n",
+                        format_type_scheme_around_name(
+                            name,
+                            &Some(inferred_scheme),
+                        )
+                    ))?;
+                }
+            }
+        }
+
+        // We have to check type schemes last, because only at this point do we know
+        // that other types that should be concrete do not occur as type variables in the
+        // inferred type scheme any more.
+        let type_var_mapping = self.inferred_types.iter().map(|(name, inferred_type)| {
+            let inferred = self.to_type_scheme(inferred_type.clone());
+            let declared = self.declared_types[name].clone().simplify_type_vars();
+            if inferred != declared {
+                Err(format!(
+                        "Inferred type scheme for symbol {name} does not match the declared type.\nInferred: let{} {name}: {}\nDeclared: let{} {name}: {}",
+                        inferred.type_vars_to_string(),
+                        inferred.ty,
+                        self.declared_types[name].type_vars_to_string(),
+                        self.declared_types[name].ty
+                    ))?;
+            }
+            let declared_type_vars = self.declared_types[name].ty.contained_type_vars();
+            let inferred_type = self.substitute_to(inferred_type.clone());
+            let inferred_type_vars = inferred_type.contained_type_vars();
+            Ok((name.clone(),
+                inferred_type_vars
+                    .into_iter()
+                    .cloned()
+                    .zip(declared_type_vars.into_iter().map(|tv| Type::TypeVar(tv.clone())))
+                    .collect(),
+            ))
+        }).collect::<Result<_, String>>()?;
+
+        Ok(type_var_mapping)
+    }
+
+    /// Updates generic arguments and literal annotations with the proper resolved types.
```

Comment:
what's a literal annotation?

Reply:
The second element of Expression::Number that specifies which type the literal has.